### PR TITLE
Initial draft of PBJ types for token state

### PR DESCRIPTION
**Pull request by Neeharika-Sompalli**
Fixes https://github.com/hashgraph/hedera-services/issues/5804
Fixes https://github.com/hashgraph/hedera-protobufs/issues/264

Adds **initial draft** of PBJ types Account, Token, Token-relation for token-service implementation.

Currently same fields in `MerkleAccount`, `MerkleToken` and `MerkleTokenRelStatus` are used in the protobuf messages. They are subject to change in the future.

### CODE REVIEW

File: services/state/token/account.proto [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+ *
+ * As with all network entities, an account has a unique entity number, which is usually given along
+ * with the network's shard and realm in the form of a shard.realm.number id.
+ */
+
+message Account {
+    /**
+     * The unique entity number of the account.
+     */
+    int64 accountNumber = 1;
```
Comment (tinker-michaelj) [inline services/state/token/account.proto:42]: `account_number`

File: services/state/token/account.proto [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+     */
+    int64 auto_renew_secs = 22;
+    /**
+     * If this account is a smart-contract, number of key-value pairs stored on the contract.
+     */
+    int32 num_contract_kv_pairs = 23;
+    /**
+     * (Optional) List of crypto allowances approved by the account.
+     */
+    repeated CryptoAllowance cryptoAllowances = 24;
```
Comment (tinker-michaelj) [inline services/state/token/account.proto:140]: `crypto_allowances`

File: services/state/token/account.proto [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    /**
+     * (Optional) List of non-fungible token allowances approved by the account.
+     */
+    repeated NftAllowance nftAllowances = 25;
```
Comment (tinker-michaelj) [inline services/state/token/account.proto:144]: `nft_allowances`

File: services/state/token/token.proto [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+ *
+ * As with all network entities, a token has a unique entity number, which is usually given along
+ * with the network's shard and realm in the form of a shard.realm.number id.
+ */
+
+message Token {
+    /**
+     * The unique entity number of this token.
+     */
+    int64 tokenNum = 1;
```
Comment (tinker-michaelj) [inline services/state/token/token.proto:43]: `token_num`

File: services/state/token/token.proto [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+     */
+    bool accounts_kyc_granted_by_default = 23;
+    /**
+     * The type of this token.
+     */
+    TokenType tokenType = 24;
+    /**
+     * The supply type of this token.
+     */
+    TokenSupplyType supplyType = 25;
```
Comment (tinker-michaelj) [inline services/state/token/token.proto:139]: `supply_type`

File: services/state/token/token_relation.proto [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+     */
+    bool kycGranted = 5;
+    /**
+     * The flag indicating if the token relationship was deleted.
+     */
+    bool deleted = 6;
+    /**
+     * The flag indicating if the token relationship was created using automatic association.
+     */
+    bool automaticAssociation = 8;
```
Comment (tinker-michaelj) [inline services/state/token/token_relation.proto:64]: `automatic_association`

File: services/state/token/token.proto (hunk 5 above)
Comment (Neeharika-Sompalli) [reply services/state/token/token.proto]: Fixed. Thanks

File: services/state/token/token.proto (hunk 4 above)
Comment (Neeharika-Sompalli) [reply services/state/token/token.proto]: Fixed.

File: services/state/token/account.proto [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+    /**
+     * (Optional) List of fungible token allowances approved by the account.
+     */
+    repeated TokenAllowance tokenAllowances = 26;
```
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply services/state/token/account.proto]: Fixed. Thanks

File: services/state/token/account.proto (hunk 3 above)
Comment (Neeharika-Sompalli) [reply services/state/token/account.proto]: Fixed. Thanks

File: services/state/token/account.proto (hunk 1 above)
Comment (Neeharika-Sompalli) [reply services/state/token/account.proto]: Fixed. Thanks

**Review by mhess-swl — APPROVED** [review]

File: services/state/token/account.proto [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+     */
+    uint64 head_token_number = 27;
+    /**
+     * The NftId of the head of the linked list from unique tokens map for the account.
+     */
+    uint64 head_nft_id = 28;
+    /**
+      * The serial number of the head NftId of the linked list from unique tokens map for the account.
+      */
+    uint64 head_nft_serial_number = 29;
```
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply services/state/token/account.proto]: Yes @xin-hedera . Thanks. Fixed

File: services/state/token/token_relation.proto [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+ */
+message TokenRelation {
+    /**
+     * The token involved in this relation.
+     */
+    uint64 token_number = 1;
+    /**
+     * The account involved in this association.
+     */
+    uint64 account_number = 2;
```
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply services/state/token/token_relation.proto]: Fixed. Thanks

File: services/state/token/token_relation.proto [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+option java_package = "com.hederahashgraph.api.proto.java";
+// <<<pbj.java_package = "com.hedera.hapi.node.state.token">>> This comment is special code for setting PBJ Compiler java package
+option java_multiple_files = true;
+/**
+ * First-draft representation of a Hedera Token Service token relationship entity in the network Merkle tree.
+ *
+ * As with all network entities, a token relationship has a unique entity number pair, which is represented
+ * with the account and the token involved in the relationship.
+ */
+message TokenRelation {
```
Comment (jsync-swirlds) [inline services/state/token/token_relation.proto:36]: Is this related to the existing TokenRelationship entity?

Comment (Neeharika-Sompalli) [reply services/state/token/token_relation.proto:36]: yes, `MerkleTokenRelStatus`

File: services/state/token/token_relation.proto [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+     */
+    int64 token_number = 1;
+    /**
+     * The account involved in this association.
+     */
+    int64 account_number = 2;
+    /**
+     * The balance of the token relationship.
+     */
+    int64 balance = 3;
```
Comment (jsync-swirlds) [inline services/state/token/token_relation.proto:48]: Do we want this to be a flat number, or would it make sense to use a TokenBalance entity?

File: services/state/token/account.proto [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+     */
+    Key key = 3;
+    /**
+     * The expiration time of the account, in seconds since the epoch.
+     */
+    int64 expiry = 4;
+    /**
+     * The balance of the account, in tiny-bars.
+     */
+    int64 hbar_balance = 5;
```
Comment (tinker-michaelj) [inline services/state/token/account.proto:59]: There's a few spots we can be more expressive about the units of a quantity...maybe here `tinybar_balance`?

File: services/state/token/account.proto [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+     */
+    int32 used_auto_associations = 12;
+    /**
+     * The number of tokens associated with the account.
+     */
+    int32 number_associations = 13;
+    /**
+     * The number of tokens with a positive balance associated with the account.
+     */
+    int32 number_positive_balances = 14;
```
Comment (tinker-michaelj) [inline services/state/token/account.proto:95]: `number_non_zero_token_balances` maybe?

File: services/state/token/account.proto [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+    /**
+     * The nonce of the account, used for Ethereum interoperability.
+     */
+    int64 ethereum_nonce = 15;
+    /**
+     * The amount of hbars staked to the account.
+     */
+    int64 staked_to_me = 16;
```
Comment (tinker-michaelj) [inline services/state/token/account.proto:103]: `whole_hbars_staked_to_me`?

**Review by tinker-michaelj — DISMISSED** [review]
All seems quite reasonable!

File: services/state/token/token.proto [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+option java_multiple_files = true;
+
+/**
+ * First-draft representation of a Hedera Token Service token entity in the network Merkle tree.
+ *
+ * As with all network entities, a token has a unique entity number, which is usually given along
+ * with the network's shard and realm in the form of a shard.realm.number id.
+ */
+
+message Token {
```
Comment (jsync-swirlds) [inline services/state/token/token.proto:39]: Is this related to the existing TokenInfo entity?

File: services/state/token/token_relation.proto (hunk 10 above)
Comment (jsync-swirlds) [reply services/state/token/token_relation.proto:36]: I guess what I'm asking, is if we're duplicating structure here, changing existing entities, or creating a new and unrelated entity for internal use?
If we're duplicating the entity, we need to be careful about unsynchronized changes to the two and should probably document in both that they're related and need to be kept in sync.
If we're changing the existing entity, we probably need to document that carefully and deprecate the other definition.
If we're creating a new (more-or-less) unrelated entity, then it should probably be documented here that this isn't a replacement or compliment to that entity.

File: services/state/token/token.proto (hunk 15 above)
Comment (Neeharika-Sompalli) [reply services/state/token/token.proto:39]: It is `MerkleToken` entity

File: services/state/token/token_relation.proto (hunk 10 above)
Comment (Neeharika-Sompalli) [reply services/state/token/token_relation.proto:36]: I think the idea is we will not have `Merkle*` entities after refactor. We will be using the PBJ types in all Maps in state.

File: services/state/token/token_relation.proto (hunk 11 above)
Comment (Neeharika-Sompalli) [reply services/state/token/token_relation.proto:48]: It is a long value [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/merkle/MerkleTokenRelStatus.java#L47).

File: services/state/token/account.proto (hunk 12 above)
Comment (Neeharika-Sompalli) [reply services/state/token/account.proto]: Yes, done. Thanks !

File: services/state/token/account.proto [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+    /**
+     * The expiration time of the account, in seconds since the epoch.
+     */
+    int64 expiry = 4;
+    /**
+     * The balance of the account, in tiny-bars.
+     */
+    int64 hbar_balance = 5;
+    /**
+     * An optional description of the topic with UTF-8 encoding up to 100 bytes.
```
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply services/state/token/account.proto]: It is enforced on the Transaction level. We throw `MEMO_TOO_LONG` if its larger than 100 bytes

**Review by xin-hedera — APPROVED** [review]

File: services/state/token/token_relation.proto (hunk 10 above)
Comment (jsync-swirlds) [reply services/state/token/token_relation.proto:36]: Thanks.  So this is replacing internal data structures, not replicating or replacing external entities.  That makes sense, and helps to understand where these fit.

**Review by jsync-swirlds — DISMISSED** [review]

**Review by tinker-michaelj — APPROVED** [review]
LGTM!

File: services/state/token/token_relation.proto [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+     */
+    bool kyc_granted = 5;
+    /**
+     * The flag indicating if the token relationship was deleted.
+     */
+    bool deleted = 6;
+    /**
+     * The flag indicating if the token relationship was created using automatic association.
+     */
+    bool automatic_association = 8;
```
Comment (nickpoorman) [inline services/state/token/token_relation.proto:64]: How come 7 is skipped and this protobuf field went directly to 8?

**Review by nickpoorman — CHANGES_REQUESTED** [review]

File: services/state/token/account.proto [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+ * with the network's shard and realm in the form of a shard.realm.number id.
+ */
+
+message Account {
+    /**
+     * The unique entity number of the account.
+     */
+    int64 account_number = 1;
+    /**
+     * The public key bytes to be used as the account's alias, if any.
```
Comment (rbair23) [inline services/state/token/account.proto:45]:
```suggestion
     * The alias to use for this account, if any.
```

Alias is no longer just key bytes, so either we put in a long description of the possible encodings, or we just keep it short. Either way works.

File: services/state/token/account.proto
Comment (rbair23) [inline services/state/token/account.proto:1]: We should move from `services/state/token` to `state/token`. Ideally we would have had `services/state` and `services/grpc` with sub folders for all the different kinds of schemas, but unfortunately we didn't and moving things around now breaks clients. And putting a mixture of state protobuf schemas and grpc protobuf schemas seems like it will become quite confusing. So I think moving `state` up to be a top level directory would be better.

Comment (rbair23) [reply services/state/token/account.proto:1]: And we should have a README that says that the `state` directory and sub directories are in preview mode and not final.

File: services/state/token/account.proto [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+
+option java_package = "com.hederahashgraph.api.proto.java";
+// <<<pbj.java_package = "com.hedera.hapi.node.state.token">>> This comment is special code for setting PBJ Compiler java package
+option java_multiple_files = true;
+
+/**
+ * First-draft representation of a Hedera Token Service account entity in the network Merkle tree.
+ *
+ * As with all network entities, an account has a unique entity number, which is usually given along
+ * with the network's shard and realm in the form of a shard.realm.number id.
```
Comment (rbair23) [inline services/state/token/account.proto:36]: Not a blocker for this PR, but we really should have better docs here. These schemas will generate Java objects that we will use in our code. So if I'm working on some code and I want to understand the Account class, I'm going to be looking at and reading that javadoc. So It would be nice to describe more what an account is, how it fits into our system, etc. At least at a high level.

File: services/state/token/account.proto [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+ */
+
+message Account {
+    /**
+     * The unique entity number of the account.
```
[the remark this answers is not in the document]
Comment (rbair23) [reply services/state/token/account.proto:41]: It might also be good to document that although we have an int64, only non-negative values are permitted.

File: services/state/token/account.proto [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+     */
+    int64 account_number = 1;
+    /**
+     * The public key bytes to be used as the account's alias, if any.
+     */
+    bytes alias = 2;
+    /**
+     * (Optional) The key to be used to sign transactions from the account, if any.
```
Comment (rbair23) [inline services/state/token/account.proto:49]: We should describe a little better as to when having a key is optional and when it is not. Most accounts require it. Only special accounts like 0.0.800 permit an empty key. Maybe we should document that.

File: services/state/token/account.proto (hunk 20 above)
[the remark this answers is not in the document]
Comment (rbair23) [reply services/state/token/account.proto:41]: Later we should debate whether the shard and realm should be included here or not. Because I'm not sure if we need them to be part of the object for state proofs or not.

File: services/state/token/account.proto [hunk 22]
```diff
@@ -0,0 +1,176 @@
...
+     */
+    Key key = 3;
+    /**
+     * The expiration time of the account, in seconds since the epoch.
+     */
+    int64 expiry = 4;
+    /**
+     * The balance of the account, in tiny-bars.
+     */
+    int64 tinybar_balance = 5;
```
Comment (rbair23) [inline services/state/token/account.proto:59]: Minor nit: should it be `tinybar_balance` or `tiny_bar_balance`? The difference will be `TinybarBalance` or `TinyBarBalance`. I think what you have is right, I was just wondering.

File: services/state/token/account.proto [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+     */
+    bool deleted = 7;
+    /**
+     * A boolean marking if the account is a smart contract.
+     */
+    bool smart_contract = 8;
+    /**
+     * A boolean marking if the account requires a receiver signature.
+     */
+    bool receiver_sig_required = 9;
```
Comment (rbair23) [inline services/state/token/account.proto:75]: These all look good. However there is one point worth making, which is that in protobuf serialization, false is represented by omitting the data from the data, whereas "true" is represented by a single byte. So wherever we can, we should make sure the "false" setting is more frequent than the "true". In these three cases, that is what we see. most accounts have `deleted` as false, most accounts are not smart contracts, and most accounts don't have `receiver_sig_required` set to true, so for most accounts these three fields will encode as 0 bytes. Which is great.

File: services/state/token/account.proto [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+    /**
+     * The number of NFTs owned by the account.
+     */
+    int64 number_owned_nfts = 10;
```
Comment (rbair23) [inline services/state/token/account.proto:79]: Is there any reason we use `int64` here, but `int32` in associations? Should we use `int32` in all these places? I'm not sure if anybody is going to have more than 2 billion NFTs, but I guess it could happen... but maybe they should use multiple accounts at that point?

At the protobuf level int64 and int32 are just as efficient up until the limit of int32. But at the Java level it is a long vs. an int. Otherwise it wouldn't matter that much.

File: services/state/token/account.proto [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+     */
+    int32 max_auto_associations = 11;
+    /**
+     * The number of used auto-association slots.
+     */
+    int32 used_auto_associations = 12;
+    /**
+     * The number of tokens associated with the account.
+     */
+    int32 number_associations = 13;
```
Comment (rbair23) [inline services/state/token/account.proto:91]: What happens if number_associations is maxed out at the int32 max value, but used_auto_associations are not maxed out, and somebody tries to then create a new auto_association? Do we error out, or do we end up overflowing on number_associations? Just queries.

Comment (rbair23) [reply services/state/token/account.proto:91]: Also, what do we use number_associations for? Is it just for a response to a query?

File: services/state/token/account.proto [hunk 26]
```diff
@@ -0,0 +1,176 @@
...
+    /**
+     * The number of tokens with a positive balance associated with the account.
+     */
+    int32 number_positive_balances = 14;
```
Comment (rbair23) [inline services/state/token/account.proto:95]: Is this only needed to answer some queries? Or do we use it during checking for solvency? Or something else?

File: services/state/token/account.proto [hunk 27]
```diff
@@ -0,0 +1,176 @@
...
+    /**
+     * The amount of hbars staked to the account at the start of the last rewarded period.
+     */
+    int64 stake_at_start_of_last_rewarded_period = 20;
+    /**
+     * (Optional) The number of an auto-renew account, in the same shard and realm as the account, that
+     * has signed a transaction allowing the network to use its balance to automatically extend the account's
+     * expiration time when it passes.
+     */
+    int64 auto_renew_account_number = 21;
```
Comment (rbair23) [inline services/state/token/account.proto:128]: If we end up with shard and realm encoded in these files, we'll need it here as well. Otherwise we can leave it off.

File: services/state/token/account.proto [hunk 28]
```diff
@@ -0,0 +1,176 @@
...
+    /**
+     * The number of seconds the network should automatically extend the account's expiration by, if the
+     * account has a valid auto-renew account, and is not deleted upon expiration.
+     */
+    int64 auto_renew_secs = 22;
```
Comment (rbair23) [inline services/state/token/account.proto:133]: What is the behavior if this is ever 0? Do we just make sure it never is?

File: services/state/token/account.proto [hunk 29]
```diff
@@ -0,0 +1,176 @@
...
+    /**
+     * If this account is a smart-contract, number of key-value pairs stored on the contract.
+     */
+    int32 contract_kv_pairs_number = 23;
```
Comment (rbair23) [inline services/state/token/account.proto:137]: This appears to be the only field that is smart contract specific (other than the smart_contract boolean). I'm wondering why we need it? Is it again for the purposes of a query?

File: services/state/token/account.proto [hunk 30]
```diff
@@ -0,0 +1,176 @@
...
+    /**
+     * (Optional) List of fungible token allowances approved by the account.
+     */
+    repeated TokenAllowance token_allowances = 26;
+    /**
+     * The token ID of the head of the linked list from token relations map for the account.
+     */
+    int64 head_token_number = 27;
+    /**
+     * The NftId of the head of the linkeint64d list from unique tokens map for the account.
```
Comment (rbair23) [inline services/state/token/account.proto:156]: `linkeint64d` looks like a typo

File: services/state/token/account.proto [hunk 31]
```diff
@@ -0,0 +1,176 @@
...
+    repeated CryptoAllowance crypto_allowances = 24;
+    /**
+     * (Optional) List of non-fungible token allowances approved by the account.
+     */
+    repeated NftAllowance nft_allowances = 25;
+
+    /**
+     * (Optional) List of fungible token allowances approved by the account.
+     */
+    repeated TokenAllowance token_allowances = 26;
```
Comment (rbair23) [inline services/state/token/account.proto:150]: I'm not sure these are right. It seems like we should be using our "map of linked list nodes" concept here instead. It isn't a blocker for this PR, but we need to discuss this.

File: services/state/token/token.proto [hunk 32]
```diff
@@ -0,0 +1,144 @@
...
+     */
+    int64 token_number = 1;
+    /**
+     * The name of this token.
+     */
+    string name = 2;
+    /**
+     * The symbol of this token.
+     */
+    string symbol = 3;
```
Comment (rbair23) [inline services/state/token/token.proto:51]: We should document max lengths

File: services/state/token/token.proto [hunk 33]
```diff
@@ -0,0 +1,144 @@
...
+     */
+    Key supply_key = 11;
+    /**
+     * (Optional) The fee schedule key of this token.
+     */
+    Key fee_schedule_key = 12;
+    /**
+     * (Optional) The pause key of this token.
+     */
+    Key pause_key = 13;
```
Comment (rbair23) [inline services/state/token/token.proto:91]: We should document the expectation for these keys -- like if a key is not set when the object is created, it cannot be set later, or only if there is an admin key, or whatever. Just so whoever is working in the code can see the docs right there, and not have to go back and look at the grpc objects to understand the invariants.

File: services/state/token/token.proto [hunk 34]
```diff
@@ -0,0 +1,144 @@
...
+     */
+    bool accounts_frozen_by_default = 22;
+    /**
+      * The flag indicating if this token has accounts associated to it that are KYC granted by default.
+     */
+    bool accounts_kyc_granted_by_default = 23;
+    /**
+     * The type of this token.
+     */
+    TokenType token_type = 24;
```
Comment (rbair23) [inline services/state/token/token.proto:135]: Ooh, I just realized, we're reusing the HAPI objects here, which is interesting. It isn't a bad thing, but I wonder if it will give the PBJ compiler fits when it is generating code, if `state` is at the top level and not under `services`...

File: services/state/token/account.proto [hunk 35]
```diff
@@ -0,0 +1,176 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ‍
+ */
+
+import "basic_types.proto";
+import "crypto_approve_allowance.proto";
+
+option java_package = "com.hederahashgraph.api.proto.java";
```
Comment (rbair23) [inline services/state/token/account.proto:28]: We should use the same package name here as we do for PBJ, just for clarity (no reason for it to be different).

File: services/state/token/account.proto [hunk 36]
```diff
@@ -0,0 +1,176 @@
...
+// <<<pbj.java_package = "com.hedera.hapi.node.state.token">>> This comment is special code for setting PBJ Compiler java package
```
Comment (rbair23) [inline services/state/token/account.proto:29]: This doesn't seem like the right package. Should it be in `com.hedera.node.app.service.token.model` or something like that.

File: services/state/token/account.proto (hunk 20 above)
[the remark this answers is not in the document]
Comment (nickpoorman) [reply services/state/token/account.proto:41]: If only non-negative numbers are permitted, should we make it a `uint32` or `uint64` and deal with it on the backend?

[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply services/state/token/account.proto]: Agreed @rbair23 . I think shard and realm can be common for a node in state or somewhere in one place.

File: services/state/token/account.proto (hunk 18 above)
Comment (Neeharika-Sompalli) [reply services/state/token/account.proto]: Makes sense. Thanks

File: services/state/token/account.proto (hunk 22 above)
Comment (Neeharika-Sompalli) [reply services/state/token/account.proto:60]: Yeah, I wondered about the same and confirmed `tinybar` is being used in sdks and other places in code🙂

**Comment by Neeharika-Sompalli** [comment]
> I'm not sure about where we should put things. Ideally I like moving it out of `services` just so we don't have confusion about states vs. HAPI/gRPC, but maybe right now we have no choice but to put it where you have it in this PR. I had a lot of comments and suggestions, but I don't want to hold up development for them right now, but they are questions we will need to have answered before we're ready for production.
> 
> 1. Where to put things?
> 2. Do we need to encode shard and realm as well?
> 3. Should we reuse HAPI objects? (If we do reuse them, it makes it almost impossible to fork these objects if we need more state or something else later)
> 4. What should the java package be for generated code? And what module runs the compiler to build it?

As per the discussion
1. We will retain them under `services/state`
2. We will need to add `shard` and `realm` to the objects. Need to check with platform team before adding them. So will leave for now on this PR.
3. We will reuse HAPI objects. But the PBJ compiler will keep generated objects under different directories. 
4. Decided on `pbj.java_package =  "com.hedera.hapi.node.state.token"` and `java_package = "com.hederahashgraph.api.proto.java.state"`

File: services/state/token/account.proto (hunk 36 above)
Comment (Neeharika-Sompalli) [reply services/state/token/account.proto:29]: Fixed as per discussion.

File: services/state/token/account.proto (hunk 35 above)
Comment (Neeharika-Sompalli) [reply services/state/token/account.proto]: Fixed as per discussion.

File: services/state/token/token.proto (hunk 34 above)
Comment (Neeharika-Sompalli) [reply services/state/token/token.proto:135]: As discussed , PBJ compiler will put under different directories i the future.

File: services/state/token/account.proto (hunk 23 above)
Comment (Neeharika-Sompalli) [reply services/state/token/account.proto:77]: Great insight. Thanks !

File: services/state/token/account.proto (hunk 24 above)
Comment (Neeharika-Sompalli) [reply services/state/token/account.proto:81]: Just used int64 where we were using long and `int32` for `int`. Will revisit before these going to production.

File: services/state/token/token.proto (hunk 32 above)
Comment (Neeharika-Sompalli) [reply services/state/token/token.proto]: Done.

File: services/state/token/account.proto (hunk 29 above)
Comment (Neeharika-Sompalli) [reply services/state/token/account.proto:139]: I think this is needed for Storage rent.

File: services/state/token/account.proto (hunk 25 above)
Comment (Neeharika-Sompalli) [reply services/state/token/account.proto:93]: Checked with @tinker-michaelj . It is not only needed for query, but also for calculating fee for renewal.

File: services/state/token/account.proto (hunk 26 above)
Comment (Neeharika-Sompalli) [reply services/state/token/account.proto:97]: It is needed because account can't be deleted if there are any positive balances in it.

File: services/state/token/account.proto (hunk 28 above)
Comment (Neeharika-Sompalli) [reply services/state/token/account.proto:135]: Yeah we are checking if auto renew seconds is in the range in precheck based on these properties
```
ledger.autoRenewPeriod.maxDuration=8000001
ledger.autoRenewPeriod.minDuration=2592000
```

File: services/state/token/account.proto (hunk 21 above)
Comment (Neeharika-Sompalli) [reply services/state/token/account.proto:51]: And hollow accounts ?

File: services/state/token/token.proto (hunk 33 above)
Comment (Neeharika-Sompalli) [reply services/state/token/token.proto]: Done

File: services/state/token/account.proto [hunk 37]
```diff
@@ -0,0 +1,185 @@
...
+     */
+    string memo = 6;
+    /**
+     * A boolean marking if the account has been deleted.
+     */
+    bool deleted = 7;
+    /**
+     * A boolean marking if the account is a smart contract.
+     */
+    bool smart_contract = 8;
```
Comment (jsync-swirlds) [inline services/state/token/account.proto:75]: reordering fields one of the few ways to break protobuf compatabilty.  Do we want to move this to an ordinal higher than 16 now?

File: services/state/token/token.proto [hunk 38]
```diff
@@ -0,0 +1,164 @@
...
+     */
+    int64 expiry = 18;
+    /**
+     * An optional description of the token with UTF-8 encoding up to 100 bytes.
+     */
+    string memo = 19;
+    /**
+     * The maximum supply of this token.
+     */
+    int64 max_supply = 20;
```
Comment (jsync-swirlds) [inline services/state/token/token.proto:137]: Should this be in the first 16 fields?

Comment (Neeharika-Sompalli) [reply services/state/token/token.proto]: I think all keys should also be in first 16 fields , so the fields are already filled in. 🙂

File: services/state/token/account.proto (hunk 37 above)
Comment (Neeharika-Sompalli) [reply services/state/token/account.proto]: yes fixed. Thanks !

**Review by jsync-swirlds — APPROVED** [review]

**Review by rbair23 — APPROVED** [review]

**Review by nickpoorman — APPROVED** [review]

**Review by xin-hedera — APPROVED** [review]
